File: app/src/main/java/com/aegamesi/steamtrade/fragments/adapters/LibraryAdapter.java
package com.aegamesi.steamtrade.fragments.adapters;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.ViewHolder;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.aegamesi.steamtrade.R;
import com.aegamesi.steamtrade.fragments.FragmentLibrary;
import com.aegamesi.steamtrade.libs.AndroidUtil;
import com.bumptech.glide.Glide;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;

public class LibraryAdapter extends RecyclerView.Adapter<LibraryAdapter.ViewHolderGame> {
	public static final int SORT_ALPHABETICAL = 0;
	public static final int SORT_PLAYTIME = 1;
	public int currentSort = SORT_ALPHABETICAL;
	private List<FragmentLibrary.LibraryEntry> games = null;
	private List<FragmentLibrary.LibraryEntry> filteredList;
	private FragmentLibrary fragment;
	private Context context;

	public LibraryAdapter(Context context, FragmentLibrary fragmentLibrary) {
		this.context = context;
		this.fragment = fragmentLibrary;
		filteredList = new ArrayList<>();
	}

	@NonNull
	@Override
	public ViewHolderGame onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
		return new ViewHolderGame(parent);
	}

	@Override
	public void onBindViewHolder(@NonNull ViewHolderGame h, int position) {
		FragmentLibrary.LibraryEntry entry = filteredList.get(position);

		h.itemView.setTag(entry);

		h.imageHeader.setImageResource(R.drawable.default_game);
		if (entry.img_logo_url.trim().length() > 0) {
			String image_url = "http://media.steampowered.com/steamcommunity/public/images/apps/%d/%s.jpg";

			Glide.with(context)
					.load(String.format(Locale.US, image_url, entry.appid, entry.img_logo_url))
					.into(h.imageHeader);
		}

		h.textName.setText(entry.name);

		String playtimeText = "";
		if (entry.playtime_forever > 0.0 && entry.playtime_2weeks == 0.0) {
			String baseString = h.itemView.getResources().getString(R.string.library_playtime);
			playtimeText = String.format(baseString, entry.playtime_forever);
		} else if (entry.playtime_forever > 0.0 && entry.playtime_2weeks > 0.0) {
			String baseString = h.itemView.getResources().getString(R.string.library_playtime_full);
			playtimeText = String.format(baseString, entry.playtime_2weeks, entry.playtime_forever);
		}
		h.textPlaytime.setText(playtimeText);
		h.textPlaytime.setVisibility(playtimeText.length() > 0 ? View.VISIBLE : View.GONE);
	}

	@Override
	public int getItemViewType(int position) {
		return 0;
	}

	@Override
	public int getItemCount() {
		return filteredList == null ? 0 : filteredList.size();
	}

	public void filter(String by) {
		filteredList.clear();
		if (games == null)
			return;
		if (by == null || by.trim().length() == 0) {
			filteredList.addAll(games);
		} else {
			for (FragmentLibrary.LibraryEntry game : games)
				if (game.name.toLowerCase(Locale.ENGLISH).contains(by.toLowerCase(Locale.ENGLISH)))
					filteredList.add(game);
		}
		notifyDataSetChanged();
	}

	public void setGames(List<FragmentLibrary.LibraryEntry> list, final int sort) {
		if (list != null) {
			Collections.sort(list, (lhs, rhs) -> {
				if (sort == SORT_ALPHABETICAL)
					return lhs.name.compareTo(rhs.name);
				if (sort == SORT_PLAYTIME)
					return -AndroidUtil.numCompare(lhs.playtime_forever, rhs.playtime_forever);
				return 0;
			});
		}
		currentSort = sort;
		games = list;
		filter("");
	}

	class ViewHolderGame extends ViewHolder {
		ImageView imageHeader;
		TextView textName;
		TextView textPlaytime;

		ViewHolderGame(ViewGroup parent) {
			super(LayoutInflater.from(parent.getContext()).inflate(R.layout.fragment_library_item, parent, false));

			imageHeader = itemView.findViewById(R.id.game_banner);
			textName = itemView.findViewById(R.id.game_name);
			textPlaytime = itemView.findViewById(R.id.game_playtime);

			itemView.setOnClickListener(fragment);
		}
	}
}

File: app/src/main/java/com/aegamesi/steamtrade/fragments/adapters/LibraryAdapter.kt
package com.aegamesi.steamtrade.fragments.adapters

import android.content.Context
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.aegamesi.steamtrade.R
import com.aegamesi.steamtrade.fragments.FragmentLibrary
import com.bumptech.glide.Glide
import java.util.ArrayList
import java.util.Locale
import kotlin.Comparator

class LibraryAdapter(private val context: Context,
                     private val fragment: FragmentLibrary) : RecyclerView.Adapter<LibraryAdapter.ViewHolderGame>() {

    var currentSort = SORT_ALPHABETICAL
    private var games: List<FragmentLibrary.LibraryEntry>? = null
    private val filteredList: MutableList<FragmentLibrary.LibraryEntry>?

    companion object {
        const val SORT_ALPHABETICAL = 0
        const val SORT_PLAYTIME = 1
    }

    init {
        filteredList = ArrayList()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolderGame {
        return ViewHolderGame(parent)
    }

    override fun onBindViewHolder(h: ViewHolderGame, position: Int) {
        val entry = filteredList!![position]

        h.itemView.tag = entry
        h.imageHeader.setImageResource(R.drawable.default_game)

        if (entry.imgLogoUrl.isNotEmpty()) {
            val imageURL = "http://media.steampowered.com/steamcommunity/public/images/apps/%d/%s.jpg"
            val formattedURL = String.format(Locale.US, imageURL, entry.appId, entry.imgLogoUrl)

            Glide.with(context)
                    .load(formattedURL)
                    .into(h.imageHeader)
        }

        h.textName.text = entry.name

        var playtimeText = ""
        if (entry.playtimeForever > 0.0 && entry.playtime2weeks == 0.0) {
            val baseString = h.itemView.resources.getString(R.string.library_playtime)
            playtimeText = String.format(baseString, entry.playtimeForever)
        } else if (entry.playtimeForever > 0.0 && entry.playtime2weeks > 0.0) {
            val baseString = h.itemView.resources.getString(R.string.library_playtime_full)
            playtimeText = String.format(baseString, entry.playtime2weeks, entry.playtimeForever)
        }
        h.textPlaytime.text = playtimeText
        h.textPlaytime.visibility = if (playtimeText.isNotEmpty()) View.VISIBLE else View.GONE
    }

    override fun getItemViewType(position: Int): Int {
        return 0
    }

    override fun getItemCount(): Int {
        return filteredList?.size ?: 0
    }

    fun filter(by: String?) {
        filteredList!!.clear()
        if (games == null)
            return

        if (by == null || by.trim { it <= ' ' }.isEmpty()) {
            filteredList.addAll(games!!)
        } else {
            for (game in games!!)
                if (game.name.toLowerCase(Locale.ENGLISH).contains(by.toLowerCase(Locale.ENGLISH)))
                    filteredList.add(game)
        }

        notifyDataSetChanged()
    }

    fun setGames(list: MutableList<FragmentLibrary.LibraryEntry>, sort: Int) {

        if(sort == SORT_ALPHABETICAL) {
            Log.d("Lib Adapter", "sort() -> l1, l2 = A-Z sort")
            list.sortWith(Comparator {
                l1,
                l2 ->
                l1.name.compareTo(l2.name)
            })
        }
        else if (sort == SORT_PLAYTIME) {
            Log.d("Lib Adapter", "sort() -> l2, l1 = 9-0 sort.")
            list.sortWith(Comparator {
                l2,
                l1 ->
                l1.playtimeForever.compareTo(l2.playtimeForever)
            })
        }

        //list?.sortWith(Comparator { _, _ ->
        //    if (sort == SORT_ALPHABETICAL)
        //        list.sortedWith(Comparator {l1, l2-> l1.name.compareTo(l2.name) })
        //    if (sort == SORT_PLAYTIME)
        //        list.sortedWith(Comparator {l2, l1-> l1.playtimeForever.compareTo(l2.playtimeForever)})
        //    0
        //})
        currentSort = sort
        games = list
        filter("")
    }

    inner class ViewHolderGame(parent: ViewGroup) : ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.fragment_library_item, parent, false)) {
        var imageHeader: ImageView = itemView.findViewById(R.id.game_banner)
        var textName: TextView = itemView.findViewById(R.id.game_name)
        var textPlaytime: TextView = itemView.findViewById(R.id.game_playtime)

        init {
            itemView.setOnClickListener(fragment)
        }
    }
}
